### PR TITLE
Add methods for SAS

### PR DESCRIPTION
**Pull request by Chethan2k1**

### CODE REVIEW

File: lib/crypto/client.cpp [hunk 1]
```diff
@@ -378,6 +378,94 @@ OlmClient::create_outbound_session(const std::string &identity_key, const std::s
...
+        auto pub_key_buffer = to_binary_buf(their_public_key);
+
+        const int ret = olm_sas_set_their_key(sas, pub_key_buffer.data(), pub_key_buffer.size());
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", sas);
+}
+
+std::vector<int>
+OlmClient::generate_bytes(OlmSAS *sas, std::string info, mtx::events::msg::SASMethods method)
```
Comment (deepbluev7) [inline lib/crypto/client.cpp:420]: Not so sure about this. I think, if you are already returning different things depending on the method, you should probably return the emojis or the number depending on the method. At which point it would already be easier to have 2 separate methods or just letting nheko do it. I think it actually may already do that partially.

File: lib/crypto/client.cpp [hunk 2]
```diff
@@ -378,6 +378,92 @@ OlmClient::create_outbound_session(const std::string &identity_key, const std::s
...
+        auto pub_key_buffer = to_binary_buf(their_public_key);
+
+        const int ret = olm_sas_set_their_key(sas, pub_key_buffer.data(), pub_key_buffer.size());
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", sas);
+}
+
+void
+OlmClient::generate_bytes_decimal(OlmSAS *sas, std::string info, std::vector<int> &output_list)
```
Comment (deepbluev7) [inline lib/crypto/client.cpp:420]: Why did you change this to an output parameter?

Comment (Chethan2k1) [reply lib/crypto/client.cpp:420]: Rather than outputting a vector I felt this was neater

Comment (deepbluev7) [reply lib/crypto/client.cpp:420]: I'm not so sure about that. Having a function return a value, when it is used to get a value, is usually the idiomatic part. Especially since it make the usage code more complicated, i.e.:

```
        std::vector<int> alice_emoji;
        alice->generate_bytes_emoji(alice_sas.get(), info, alice_emoji);
```

vs

```
        std::vector<int> alice_emoji = alice->generate_bytes_emoji(alice_sas.get(), info);
```

Comment (Chethan2k1) [reply lib/crypto/client.cpp:420]: Yeah actually, I'll change it then

File: include/mtxclient/crypto/client.hpp [hunk 3]
```diff
@@ -188,6 +194,14 @@ class OlmClient : public std::enable_shared_from_this<OlmClient>
         OlmAccount *account() { return account_.get(); }
         OlmUtility *utility() { return utility_.get(); }
 
+        //! SAS related stuff
+        //! this creates a unique pointer of class OlmSAS
+        SASPtr sas_init();
+        std::string sas_get_pub_key(OlmSAS *sas);
+        void set_their_key(OlmSAS *sas, std::string their_public_key);
+        std::vector<int> generate_bytes_decimal(OlmSAS *sas, std::string info);
+        std::vector<int> generate_bytes_emoji(OlmSAS *sas, std::string info);
```
Comment (deepbluev7) [inline include/mtxclient/crypto/client.hpp:203]: I think it may be cleaner to wrap this in a custom class. This could maybe look like this:

```
struct SAS
{
        SAS(std::string their_public_key); // constructor, does init and set_their_key
        std::string public_key(); // returns public key
        std::vector<int> generate_bytes_decimal(OlmSAS *sas, std::string info);
        std::vector<int> generate_bytes_emoji(OlmSAS *sas, std::string info);
private:
        SASPtr sas;
};

Comment (Chethan2k1) [reply include/mtxclient/crypto/client.hpp:203]: Okay

Comment (Chethan2k1) [reply include/mtxclient/crypto/client.hpp:203]: But ig in the constructor I wont be taking ```their_public_key``` as parameter cause when I use it ```nheko``` I would get their public key in the later stages.

Comment (deepbluev7) [reply include/mtxclient/crypto/client.hpp:203]: Well, I'm not sure, how you intend to use it, but I thought you would only allocate the SAS object, when you actually want to use it and you may have the public key available in that case already. If that doesn't match with what you intend to do, then yes, taking the public key separately makes sense :3

Comment (Chethan2k1) [reply include/mtxclient/crypto/client.hpp:203]: I was planning to create a SAS object for every flow and then when I get the ```key.verification.start``` that contains the public key of the other user I would assign it then.

Comment (deepbluev7) [reply include/mtxclient/crypto/client.hpp:203]: Yeah, makes sense. You also need to set the key alter for your tests.

**Comment by Chethan2k1** [comment]
I'm yet to add some stuff related to ```mac``` until then don't merge this branch.

**Comment by deepbluev7** [comment]
I converted this PR to a draft, so that it doesn't get merged. Just mark it as "Ready fro Review" once you think it is ready.

File: include/mtxclient/crypto/client.hpp [hunk 4]
```diff
@@ -85,13 +90,27 @@ unpickle(const std::string &pickled, const std::string &key)
...
 {
         BinaryBuf data;
         uint32_t message_index;
 };
 
+struct SAS
+{
+        SAS();
+        std::string public_key();
+        void set_their_key(std::string their_public_key);
```
Comment (deepbluev7) [inline include/mtxclient/crypto/client.hpp:105]: Could you maybe write a short doc string for each of those explaining how they are meant to be used, i.e.

```cpp
//! After constructing a SAS object, you can set the key for the other party using this function. This needs to be done before generating bytes or calculating a MAC.
```

With that I'd say this should be ready to merge, so you may want to undraft it?

File: lib/crypto/client.cpp [hunk 5]
```diff
@@ -378,6 +378,126 @@ OlmClient::create_outbound_session(const std::string &identity_key, const std::s
...
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", this->sas.get());
+}
+
+std::vector<int>
+SAS::generate_bytes_decimal(std::string info)
+{
+        auto input_info_buffer = to_binary_buf(info);
+        auto output_buffer     = create_buffer(5);
```
Comment (deepbluev7) [inline lib/crypto/client.cpp:428]: iirc, create_buffer creates a buffer filled with random numbers. You may want to call BinaryBuf(5) instead, I think, if you are not creating a cryptographic key but are just using the buffer as a resource.

File: include/mtxclient/crypto/client.hpp (hunk 4 above)
Comment (Chethan2k1) [reply include/mtxclient/crypto/client.hpp:105]: Yeah I'll resolve that issue and add some explanation, maybe then if nothing else is left to do I'll undraft it
